crates/astria-composer/src/searcher/executor.rs

```diff
@@ -120,178 +139,266 @@ impl Executor {
+/// Queries the sequencer for the latest nonce with an exponential backoff
+#[instrument(
+    name = "submit signed transaction",
+    skip_all,
+    fields(
+        nonce = tx.unsigned_transaction().nonce,
+        transaction.hash = hex::encode(tx.to_raw().encode_to_vec()),
+    )
+)]
+async fn submit_tx(
+    client: sequencer_client::HttpClient,
+    tx: SignedTransaction,
+) -> eyre::Result<tx_sync::Response> {
+    debug!("submitting signed transaction to sequencer");
+    let span = Span::current();
+    let retry_config = tryhard::RetryFutureConfig::new(1024)
+        .exponential_backoff(Duration::from_millis(200))
+        .max_delay(Duration::from_secs(60))
+        .on_retry(
+            |attempt,
+             next_delay: Option<Duration>,
+             err: &sequencer_client::extension_trait::Error| {
+                let wait_duration = next_delay
+                    .map(humantime::format_duration)
+                    .map(tracing::field::display);
+                let err = err.clone();
+                let span = warn_span!(parent: span.clone(), "report attempt failure");
+                async move {
+                    let error = &err as &(dyn std::error::Error + 'static);
+                    warn!(
+                        attempt,
+                        wait_duration,
+                        error,
+                        "failed getting latest nonce from sequencer; retrying after backoff",
+                    );
+                }
+                .instrument(span)
+            },
         );
+    tryhard::retry_fn(|| {
+        let client = client.clone();
+        let tx = tx.clone();
+        let span = info_span!(parent: span.clone(), "attempt submission");
+        async move { client.submit_transaction_sync(tx).await }.instrument(span)
+    })
+    .with_config(retry_config)
+    .await
+    .wrap_err("failed submitting a signed transaction after 1024 attempts")
+}
 
-        self.status.send_modify(|status| {
-            status.is_connected = true;
-        });
-        Ok(())
+// TODO: zeroize this
+pin_project! {
+    /// A future to submit a bundle to the sequencer, returning the next nonce that should be used for the next submission.
+    ///
+    /// The future will fetch a new nonce from the sequencer if a submission returned an `INVALID_NONCE` error code.
+    ///
+    /// The future will only return an error if it ultimately failed submitting a transaction due to the underlying
+    /// transport failing. This can be taken as a break condition to exit the executor loop.
+    ///
+    /// If the sequencer returned a non-zero abci code (albeit not `INVALID_NONCE`), this future will return with
+    /// that nonce it used to submit the non-zero abci code request.
+    struct SubmitFut {
+        client: sequencer_client::HttpClient,
+        address: Address,
+        nonce: u32,
+        signing_key: SigningKey,
+        #[pin]
+        state: SubmitState,
```

Comment:
is this the only field that needs to be pinned b/c it contains pinned futures?

Reply:
Yeah. All other fields are either Unpin or don't need pinning because they are not mutated.